src/integrations.py

```diff
@@ -0,0 +1,233 @@
+from charms.oathkeeper.v0.oathkeeper_info import OathkeeperInfoRequirer
+from charms.openfga_k8s.v1.openfga import OpenFGARequires
+from charms.tempo_k8s.v0.tracing import TracingEndpointRequirer
+from ops.model import Model
+
+from constants import PEER_INTEGRATION_NAME
+from env_vars import EnvVars
+
+logger = logging.getLogger(__name__)
+
+
+class PeerData:
+    def __init__(self, model: Model) -> None:
+        self._model = model
+        self._app = model.app
+
+    def __getitem__(self, key: str) -> dict:
+        if not (peers := self._model.get_relation(PEER_INTEGRATION_NAME)):
+            return {}
+
+        value = peers.data[self._app].get(key)
+        return json.loads(value) if value else {}
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if not (peers := self._model.get_relation(PEER_INTEGRATION_NAME)):
+            return
+
+        peers.data[self._app][key] = json.dumps(value)
+
+    def pop(self, key: str) -> dict:
+        if not (peers := self._model.get_relation(PEER_INTEGRATION_NAME)):
+            return {}
+
+        data = peers.data[self._app].pop(key, None)
+        return json.loads(data) if data else {}
+
+
+@dataclass(frozen=True)
+class OpenFGAModelData:
+    model_id: str = ""
+
+    def to_env_vars(self) -> EnvVars:
+        return {
+            "OPENFGA_AUTHORIZATION_MODEL_ID": self.model_id,
+        }
+
+    @classmethod
+    def load(cls, source: Mapping[str, Any]) -> "OpenFGAModelData":
+        return OpenFGAModelData(
+            model_id=source.get("openfga_model_id", ""),
+        )
+
+
+@dataclass(frozen=True)
+class OpenFGAIntegrationData:
+    url: str = ""
+    api_token: str = ""
+    store_id: str = ""
+
+    @property
+    def api_scheme(self) -> str:
+        return urlparse(self.url).scheme
+
+    @property
+    def api_host(self) -> str:
+        return urlparse(self.url).netloc
+
+    def to_env_vars(self) -> EnvVars:
+        return {
+            "OPENFGA_STORE_ID": self.store_id,
+            "OPENFGA_API_TOKEN": self.api_token,
+            "OPENFGA_API_SCHEME": self.api_scheme,
+            "OPENFGA_API_HOST": self.api_host,
+        }
+
+
+class OpenFGAIntegration:
+    def __init__(self, integration_requirer: OpenFGARequires) -> None:
+        self._openfga_requirer = integration_requirer
+
+    def is_store_ready(self) -> bool:
+        provider_data = self._openfga_requirer.get_store_info()
+        return provider_data is not None and provider_data.store_id is not None
+
+    @property
+    def openfga_integration_data(self) -> OpenFGAIntegrationData:
+        if not (provider_data := self._openfga_requirer.get_store_info()):
+            return OpenFGAIntegrationData()
+
+        return OpenFGAIntegrationData(
+            url=provider_data.http_api_url,  # type: ignore[arg-type]
+            api_token=provider_data.token,  # type: ignore[arg-type]
+            store_id=provider_data.store_id,  # type: ignore[arg-type]
+        )
```

Comment:
I am not sure about how I feel about the integration abstraction. While it does provide a separation of logic, which is needed, I am not sure about the extra complexity it adds to the project. I will try to write down some thoughts, but please take them with a grain of salt as I have not really settled on how I feel about it.

AFAICT the `*Integration` objects converts the data coming from the lib to a `*Data` object and then the `*Data` object gets translated to env vars. This means that the `*Integration` object is generic, while the `*Data` object includes logic that is application specific (env vars are specific to this application version). Intuitively I did not expect this, I would think that the integration layer would be generic and not specific to the application. 

The `*Data` objects model the integration's databag. Most integration libraries come with a `dataclass` or `pydantic Model` that serves the same purpose. I understand that the purpose is to abstract the library, but this may result in double the work. For example if we were to refactor the openfga library, we would have to refactor the library and then we would have to refactor the `OpenFGAIntegrationData` class. Ofc even if we didn't have `OpenFGAIntegrationData` we would need to refactor how the data returned from the openfga library is mapped to env vars.  I understand that the added development time is minute, but I am not sure if I see the value added.

Reply:
Understood. Some of the integrations seem to be pedantic in this charm because they are relatively simple.

I would like to express the idea that integrations are a data source or sink handler that can provide charm.py with either "prepared" integration data, or a simple way to update integrations.

We can think of them as places where

- provides/flushes integration data (either source or sink)
- hide the complexity of handling the raw integration data

Some integration libs ("requirer" or "provider") provide simple enough APIs with well-designed data, still limited to just providing/updating that specific data.

From what we can observe, sometimes it involves several data sources (integrations) and processes those data in a single event callback method. This leads to multiple steps before we can either update the workload service or update other integrations.

Or we can think in this way about these integrations:

data from integrations or charm configs (by integration lib APIs or charm APIs) -> multiple integration classes (data processing units) -> processed data (goes to integration or workload)

Currently, these data process logics are usually piled somewhere in the `charm.py` script.

We can chat about this further because I am not sure if I express the idea clearly :)